cadence/contracts/services/InterchainTokenService.cdc

```diff
@@ -0,0 +1,324 @@
+            let adminCap <- contract!.getAdminCapability()
+            let minterCap <- adminCap.createNewMinter()
+            let burnerCap <- adminCap.createNewBurner()
+
+            InterchainTokenService.account.save(<-adminCap, to: InterchainTokenService.getAdminCapabilityStoragePath(self.contractAddress)!)
+            InterchainTokenService.account.save(<-minterCap, to: InterchainTokenService.getMinterCapabilityStoragePath(self.contractAddress)!)
+            InterchainTokenService.account.save(<-burnerCap, to: InterchainTokenService.getBurnerCapabilityStoragePath(self.contractAddress)!)
+            self.adminCapability = InterchainTokenService.getAdminCapabilityStoragePath(self.contractAddress)!
+            self.minterCapability = InterchainTokenService.getMinterCapabilityStoragePath(self.contractAddress)!
+            self.burnerCapability = InterchainTokenService.getBurnerCapabilityStoragePath(self.contractAddress)!
+        }
+
+    }
+
+    access(all) event TokenContractLaunched(
+        contractName: String,
+        contractAddress: Address,
+        tokenName: String,
+        tokenSymbol: String,
+    )
+
+    access(all) event InterchainTransfer(
+        tokenAddress: Address,
+        tokenName: String,
+        sourceAddress: Address,
+        destinationChain: String,
+        amount: UFix64,
+        datahash: [UInt8]
+    )
+
+    access(all) event InterchainTransferReceived(
+        commandId: String,
+        tokenId: String,
+        sourceChain: String,
+        sourceAddress: String,
+        destinationAddress: Address,
+        amount: UFix64,
+        datahash: String
+    )
+
+    access(all) event TokenManagerDeploymentStarted(
+        tokenId: String,
+        destinationChain: String,
+        //TokenManagerType??
+        params: String,
+    )
+
+    init(publicKey: String, accountCreationFee: UFix64){
+        self.accountCreationFee = accountCreationFee
+        self.tokenPublicKey = publicKey
+        self.tokens = {}
+        self.nativeTokens <- {}
+        self.managedTokens <- {}
+        self.LATEST_METADATA_VERSION = 0
+        self.INTERCHAIN_TRANSFER = Crypto.hash("INTERCHAIN_TRANSFER".utf8, algorithm: HashAlgorithm.KECCAK_256)
+        self.DEPLOY_INTERCHAIN_TOKEN = Crypto.hash("DEPLOY_INTERCHAIN_TOKEN".utf8, algorithm: HashAlgorithm.KECCAK_256)
+        self.prefixAuthCapabilityName = "TokenAuthCapability_"
+        self.prefixNativeTokenName = "NativeToken_"
+        self.prefixManagedTokenName = "ManagedToken_"
+        self.prefixVaultName = "Vault_"
+    }
+
+    access(self) fun changeAccountCreationFee(newFee: UFix64){
+        self.accountCreationFee = newFee
+    }
+
+    // public interface for initiating interhcain transfer
+   access(all) fun interchainTransfer(senderIdentity: Capability<&{AxelarGateway.SenderIdentity}>, contractName: String, contractAddress: Address, destinationChain: String, destinationAddress: String, vault: @FungibleToken.Vault, metadata: [UInt8]){
```

Comment:
Looks like this method requires a user to onboard a Flow-native token before interchain transfers can be initiated, but I think that's not an ideal user flow and can be abstracted into the contract.

From within `_takeToken` we can first check if the vault is Flow-native and hasn't been onboarded - basically that the ITS didn't deploy its defining contract - we can then extract the contract name and address using `StringUtils` methods on  `vault.getType().identifier` and run  `_onboardNativeFungibleToken` before taking the token.

That will allow the contract to conditionally onboard a native token when the user initiates an interchain transfer without the need to first query if the vault has been onboarded.

Reply:
Unfortunately, I dont think we can onboard other tokens on the fly like that. Based on the original design of the ITS https://github.com/axelarnetwork/interchain-token-service/blob/main/DESIGN.md onboarding chain native tokens should be kept seperate from the first interchain transfer due to the potential cost which might come with launching new contract on other chains.